oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/MissingBcSweeper2.java

```diff
@@ -0,0 +1,281 @@
+
+    private static final long LOGINTERVALMS = TimeUnit.MINUTES.toMillis(1);
+
+    private final RevisionContext context;
+
+    private final CommitValueResolver commitValueResolver;
+
+    private final int executingClusterId;
+
+    private final List<Integer> includedClusterIds;
+
+    private final RevisionVector headRevision;
+
+    private long totalCount;
+    private long lastCount;
+    private long startOfScan;
+    private long lastLog;
+
+    /**
+     * Creates a new sweeper v2 for the given context..
+     *
+     * @param context the revision context.
+     */
+    MissingBcSweeper2(RevisionContext context,
+                    CommitValueResolver commitValueResolver,
+                    List<Integer> includedClusterIds) {
+        this.context = checkNotNull(context);
+        this.commitValueResolver = checkNotNull(commitValueResolver);
+        this.executingClusterId = context.getClusterId();
+        this.includedClusterIds = includedClusterIds == null ? new LinkedList<>() : Collections.unmodifiableList(includedClusterIds);
+        this.headRevision= context.getHeadRevision();
+    }
+
+    /**
+     * Performs a sweep2 and reports the required updates to the given sweep
+     * listener.
+     *
+     * @param documents the documents to sweep
+     * @param listener the listener to receive required sweep update operations.
+     * @throws DocumentStoreException if reading from the store or writing to
+     *          the store failed.
+     */
+    void sweep2(@NotNull Iterable<NodeDocument> documents,
+                   @NotNull NodeDocumentSweepListener listener)
+            throws DocumentStoreException {
+        performSweep2(documents, checkNotNull(listener));
+    }
+
+    //----------------------------< internal >----------------------------------
+
+    @Nullable
+    private void performSweep2(Iterable<NodeDocument> documents,
+                                  NodeDocumentSweepListener listener)
+            throws DocumentStoreException {
+        totalCount = 0;
+        lastCount = 0;
+        startOfScan = context.getClock().getTime();
+        lastLog = startOfScan;
+
+        Iterable<Map.Entry<Path, UpdateOp>> ops = sweepOperations(documents);
+        for (List<Map.Entry<Path, UpdateOp>> batch : partition(ops, INVALIDATE_BATCH_SIZE)) {
+            Map<Path, UpdateOp> updates = newHashMap();
+            for (Map.Entry<Path, UpdateOp> entry : batch) {
+                updates.put(entry.getKey(), entry.getValue());
+            }
+            listener.sweepUpdate(updates);
+        }
+        LOG.debug("Document sweep2 finished");
+    }
+
+    private Iterable<Map.Entry<Path, UpdateOp>> sweepOperations(
+            final Iterable<NodeDocument> docs) {
+        return filter(transform(docs,
+                new Function<NodeDocument, Map.Entry<Path, UpdateOp>>() {
+
+            int yieldCnt = 0;
+            long lastYield = System.currentTimeMillis();
```

Comment:
I would rather use the `Clock` from the `RevisionContext`. It will make it easier to control the time in a unit test.

Reply:
Agreed, done now in https://github.com/apache/jackrabbit-oak/pull/247/commits/86e3a7d309fad88c349d725b4a7f8434c766dbeb